src/plugins/dashboard/public/services/dashboard_content_management_service/lib/save_dashboard_state.ts

```diff
@@ -55,121 +28,14 @@ export const saveDashboardState = async ({
-  if (saveOptions.saveAsCopy) {
-    const { panels: newPanels, references: newPanelReferences } = generateNewPanelIds(
-      panels,
-      panelReferences
-    );
-    panels = newPanels;
-    prefixedPanelReferences = newPanelReferences;
-    //
-    // do not need to generate new ids for controls.
-    // ControlGroup Component is keyed on dashboard id so changing dashboard id mounts new ControlGroup Component.
-    //
-  }
-
-  const { searchSource, searchSourceReferences } = await (async () => {
-    const searchSourceFields = await dataSearchService.searchSource.create();
-    searchSourceFields.setField(
-      'filter', // save only unpinned filters
-      filters.filter((filter) => !isFilterPinned(filter))
-    );
-    searchSourceFields.setField('query', query);
-
-    const rawSearchSourceFields = searchSourceFields.getSerializedFields();
-    const [fields, references] = extractSearchSourceReferences(rawSearchSourceFields) as [
-      DashboardSearchSource,
-      SavedObjectReference[]
-    ];
-    return { searchSourceReferences: references, searchSource: fields };
-  })();
-
-  const options = {
-    useMargins,
-    syncColors,
-    syncCursor,
-    syncTooltips,
-    hidePanelTitles,
-  };
-  const savedPanels = convertPanelMapToPanelsArray(panels, true);
-
-  /**
-   * Parse global time filter settings
-   */
-  const { from, to } = timefilter.getTime();
-  const timeFrom = timeRestore ? convertTimeToUTCString(from) : undefined;
-  const timeTo = timeRestore ? convertTimeToUTCString(to) : undefined;
-  const refreshInterval = timeRestore
-    ? (pick(timefilter.getRefreshInterval(), [
-        'display',
-        'pause',
-        'section',
-        'value',
-      ]) as RefreshInterval)
-    : undefined;
-
-  const rawDashboardAttributes: DashboardAttributes = {
-    version: convertDashboardVersionToNumber(LATEST_DASHBOARD_CONTAINER_VERSION),
-    controlGroupInput: controlGroupInput as DashboardAttributes['controlGroupInput'],
-    kibanaSavedObjectMeta: { searchSource },
-    description: description ?? '',
-    refreshInterval,
-    timeRestore,
-    options,
-    panels: savedPanels,
-    timeFrom,
-    title,
-    timeTo,
-  };
-
-  /**
-   * Extract references from raw attributes and tags into the references array.
-   */
-  const { attributes, references: dashboardReferences } = extractReferences(
-    {
-      attributes: rawDashboardAttributes,
-      references: searchSourceReferences,
-    },
-    { embeddablePersistableStateService: embeddableService }
-  );
-
-  const savedObjectsTaggingApi = savedObjectsTaggingService?.getTaggingApi();
-  const references = savedObjectsTaggingApi?.ui.updateTagsReferences
-    ? savedObjectsTaggingApi?.ui.updateTagsReferences(dashboardReferences, tags)
-    : dashboardReferences;
-
-  const allReferences = [
-    ...references,
-    ...(prefixedPanelReferences ?? []),
-    ...(controlGroupReferences ?? []),
-  ];
+  const { attributes, references } = await getDashboardState({
```

Comment:
Should save methods be updated to just take output of `getDashboardState` instead of calling `getDashboardState` internally.

I think it would be better if `getDashboardState` lived in `dashboard_app` folder and any place that needed would just get passed `{ attributes: DashboardAttributes; references: SavedObjectReference[]; }`

Reply:
I like this idea a lot, but I think I would like to scope that in a different PR. Making those changes in this PR would greatly increase the number of changes required for review.